tests/components/smarttub/test_climate.py

```diff
@@ -0,0 +1,83 @@
+"""Test the SmartTub climate platform."""
+from unittest.mock import Mock
+
+import pytest
+
+from homeassistant.components.climate.const import (
+    CURRENT_HVAC_HEAT,
+    CURRENT_HVAC_IDLE,
+    HVAC_MODE_HEAT,
+)
+from homeassistant.components.smarttub.climate import (
+    SmartTubThermostat,
+    async_setup_entry,
+)
+from homeassistant.components.smarttub.const import (
+    DEFAULT_MAX_TEMP,
+    DEFAULT_MIN_TEMP,
+    DOMAIN,
+    SMARTTUB_CONTROLLER,
+)
+from homeassistant.components.smarttub.controller import SmartTubController
+from homeassistant.const import ATTR_TEMPERATURE, TEMP_CELSIUS
+
+from tests.async_mock import create_autospec
+from tests.common import MockConfigEntry
+
+
+@pytest.fixture(name="controller")
+async def mock_controller(hass, coordinator):
+    """Mock the controller for testing."""
+
+    controller = create_autospec(SmartTubController, instance=True)
+    controller.coordinator = coordinator
+    return controller
+
+
+async def test_async_setup_entry(hass, controller, spa):
+    """Test async_setup_entry."""
+
+    entry = MockConfigEntry(unique_id="ceid1")
+    async_add_entities = Mock()
+    hass.data[DOMAIN] = {
+        entry.unique_id: {SMARTTUB_CONTROLLER: controller},
+    }
+    controller.spas = [spa]
+
+    ret = await async_setup_entry(hass, entry, async_add_entities)
+
+    assert ret is True
+    async_add_entities.assert_called()
+
+
+async def test_thermostat(coordinator, spa):
+    """Test the thermostat entity."""
+
+    coordinator.data = {
+        spa.id: {
+            "status": {
+                "heater": "ON",
+                "water": {
+                    "temperature": 38,
+                },
+                "setTemperature": 39,
+            }
+        }
+    }
+    thermostat = SmartTubThermostat(coordinator, spa)
```

Comment:
Testing the object directly might be problematic if you're relying on lifecycle events (which I believe you are).

Reply:
If you mean these: https://developers.home-assistant.io/docs/core/entity/#lifecycle-hooks

I am not using them. If you want this to work differently, let me know what I should do instead.

Reply:
You are, but not necessarily directly. `CoordinatorEntity`, for example is using `async_added_to_hass` (you would have done that manually if you weren't inheriting from it).
The point is, entities are not POPOs, so I'm not sure testing them in isolation is a good idea.
I might be wrong though, so if you want to keep this as is, and let a member of the core team chime in once they review that's totally fine.

Reply:
I see. If there is a preferred way to do this, I'm happy to change it. What do you recommend?

Should I set up entity tests using a config entry and `async_forward_entry_setup`? Something else?

Can you show me a good example in the code base of how to do the test setup?

Reply:
Well, you can look into any `test_climate.py` to find examples, but I'm not familiar with any of them, so I don't know which ones do a good job.
You can also look at the Risco integration for a more general example of testing platforms.

Reply:
I've looked, but each module seems to do something different, and I'm having trouble implementing the approach described in https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

If I do this:

```
-    thermostat = SmartTubThermostat(coordinator, spa)
+    config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(config_entry.entry_id)
+    await hass.async_block_till_done()
+    entity_id = f"climate.{spa.brand}_{spa.model}_thermostat"
+    state = hass.states.get(entity_id)
```

I can access the initial states and check them. But I don't know how to trigger an update so that I can check that the entity state changes when the underlying device changes state. If I can't do that, then I have to write a separate test for every state I want to test.

Reply:
I've tried:

- `await hass.helpers.entity_component.async_update_entity(entity_id)`
- `await coordinator.async_refresh()`
- various incantations of `hass.async_block_till_done()`

...

Reply:
Sounds like you're in the right direction. Want to post something so that I can take a look and try to understand why it doesn't work?

Reply:
https://gist.github.com/mdz/2dc4a9333abcdf6a710b40361831804b

Reply:
You're not making any changes between the two times you're getting the state, so not sure why you're expecting it to change.

Reply:
LOL, I lost `spa.get_status.return_value["heater"] = "OFF"` when reorganizing the code. It does work.